README.md
tfs
=======

![logo](https://raw.github.com/fengmk2/tfs/master/logo.png)

[TFS: Taobao FileSystem](http://code.taobao.org/p/tfs/src/) nodejs client.

* jscoverage: [98%](http://fengmk2.github.com/coverage/tfs.html)

## Install

```bash
$ npm install tfs
```

## Usage

```js
var tfs = require('tfs');

var client = tfs.createClient({
  rootServer: '$host:port',
  appkey: '$your_appkey',
});

// upload normal file
client.upload(filepath, function (err, info) {
  console.log(info);
  // { 
  //   filename: 'T2xRETBgZv1RCvBVdK.jpg', 
  //   url: 'http://img1.tfs.com/tfscom/T2xRETBgZv1RCvBVdK.jpg',
  //   size: 1024
  // }
});

// upload custom name file
client.uploadFile(filepath, '320', 'foo.jpg', function (err, info) {
  console.log(info);
  // { 
  //   filename: 'L1/1/320/foo.jpg',
  //   url: 'http://img1.tfs.com/L1/1/320/foo.jpg',
  //   size: 1984
  // }
});
```

## API

```js
/**
 * Create TFS RESTFul client.
 * 
 * @param {Object} options
 *  - {String} appkey
 *  - {String} [appLocation], default is 'tfscom'.
 *  - {Number} [uploadTimeout], upload max timeout, default is 60s.
 *  - {String} [rootServer], 'host:port' format, default is 'restful-store.vip.tbsite.net:3800'.
 *  - {Array} [imageServers], default is CDN online servers list.
 * @return {Client}
 */
function createClient(options);

/**
 * Upload a file.

 * @param {String} filepath
 * @param {Function(err, info)} callback
 *  - {Object} info
 *   - {String} name, tfs file name
 *   - {String} url, CDN url
 *   - {Number} size, file size
 * @param {Number} timeout, default is `client.uploadTimeout`.
 */
Client.prototype.upload = function (filename, callback, timeout);

/**
 * Remove a file by name.
 * 
 * @param {String} name
 * @param {Object} [options]
 *  - {Number|String} hide, 1: hidden, 0: show
 * @param {Function(err, success)} callback
 * @param {Number} timeout, default is `client.uploadTimeout`.
 */
Client.prototype.remove = function (name, options, callback, timeout);

/**
 * Download a file by name.
 * 
 * @param {String} name
 * @param {String|WriteStream} savefile, save file path or writable stream.
 * @param {Object} [options]
 *  - {Number} offset
 *  - {Number} size
 * @param {Function(err, success)} callback
 * @param {Number} timeout, default is `client.uploadTimeout`.
 */
Client.prototype.download = function (name, savefile, options, callback, timeout);

/**
 * Get file meta.
 * @param {String} name
 * @param {Object} [options]
 *  - {Number} type, 0: normal mode, 1: force mode, get meta info even it was deleted.
 * @param {Function(err, meta)} callback
 * @param {Number} timeout
 */
Client.prototype.getMeta = function (name, options, callback, timeout);

/**
 * Upload a file with custom filename.
 * 
 * @param {String} filepath
 * @param {String} uid, user id
 * @param {String} filename
 * @param {Object} [options]
 *  - {Number} offset
 * @param {Function(err, info)} callback
 * @param {Number} [timeout]
 */
Client.prototype.uploadFile = function (filepath, uid, filename, options, callback, timeout);

/**
 * Remove a custom name file.
 * 
 * @param {String} uid
 * @param {String} filename
 * @param {Function(err, success)} callback
 * @param {Number} [timeout]
 */
Client.prototype.removeFile = function (uid, filename, callback, timeout);
```

## Authors

```bash
$ git summary 

 project  : tfs
 repo age : 2 weeks
 active   : 5 days
 commits  : 23
 files    : 15
 authors  : 
    19  苏千                  82.6%
     4  fengmk2                 17.4%
```

## License 

(The MIT License)

Copyright (c) 2012 fengmk2 &lt;fengmk2@gmail.com&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

History.md

0.1.0 / 2012-12-14 
==================

  * support getMeta()
  * fixed download test cases
  * support download() file
  * add MIT License
  * publish to github

0.0.4 / 2012-12-01 
==================

  * support remove() file by name

0.0.3 / 2012-11-28 
==================

  * add refreshError listen
  * add demo
  * rename tfs.js to client.js
  * support removeFile
  * coverage getAppid()
  * add uploadFile()
  * update readme

0.0.2 / 2012-11-28 
==================

  * rename appname to appLocation
  * change report to tap
  * fixed repository url

0.0.1 / 2012-11-28 
==================

  * upload(filepath, callback)
  * 完善单元测试
  * init


test/tfs.test.js
/*!
 * tfs - test/tfs.test.js
 * Copyright(c) 2012 fengmk2 <fengmk2@gmail.com>
 * MIT Licensed
 */

"use strict";

/**
 * Module dependencies.
 */

var path = require('path');
var tfs = require('../');
var should = require('should');
var mm = require('mm');
var pedding = require('pedding');


describe('tfs.test.js', function () {

  var tfsClient = tfs.createClient({
    appkey: 'tfscom',
    rootServer: '10.232.4.44:3800',
    imageServers: [
      'img01.daily.taobaocdn.net',
      'img02.daily.taobaocdn.net',
    ],
  });

  var logopath = path.join(path.dirname(__dirname), 'logo.png');

  afterEach(function () {
    mm.restore();
  });

  it('should throw missing appkey error', function () {
    (function () {
      tfs.createClient();
    }).should.throw('missing appkey');
  });

  describe('upload()', function () {
    var client = tfs.createClient({
      appkey: 'tfscom',
      rootServer: '10.232.4.44:3800',
      imageServers: [
        'img01.daily.taobaocdn.net',
        'img02.daily.taobaocdn.net',
      ],
    });

    it('should upload logo.png to tfs', function (done) {
      client.upload(logopath, function (err, info) {
        client.appid.should.equal('1');
        should.not.exist(err);
        info.should.have.keys('name', 'url', 'size');
        info.size.should.be.a('number');
        info.name.should.be.a('string').with.match(/\.png$/);
        info.url.should.be.a('string').with.match(/\.png$/);
        info.url.should.include('http://img01.daily.taobaocdn.net/tfscom/');

        client.refreshCounter = 1;
        // next upload will refresh servers
        client.upload(logopath, function (err, info) {
          should.not.exist(err);
          info.should.have.keys('name', 'url', 'size');
          info.size.should.be.a('number');
          info.name.should.be.a('string').with.match(/\.png$/);
          info.url.should.be.a('string').with.match(/\.png$/);
          info.url.should.include('http://img02.daily.taobaocdn.net/tfscom/');

          client.upload(logopath, function (err, info) {
            should.not.exist(err);
            info.should.have.keys('name', 'url', 'size');
            info.size.should.be.a('number');
            info.name.should.be.a('string').with.match(/\.png$/);
            info.url.should.be.a('string').with.match(/\.png$/);
            info.url.should.include('http://img01.daily.taobaocdn.net/tfscom/');
            done();
          });
        });
      });
    });

    it('should upload error', function (done) {
      mm.http.requestError(/\/v1\/tfscom/, 'mock request() error');
      client.upload(logopath, function (err, info) {
        should.exist(err);
        err.message.should.equal('mock request() error');
        should.not.exist(info);
        done();
      });
    });

    it('should upload 500 error', function (done) {
      mm.http.request(/\/v1\/tfscom/, new Buffer(''), {});
      client.upload(logopath, function (err, info) {
        should.exist(err);
        err.message.should.equal('Unexpected end of input');
        should.not.exist(info);
        done();
      });
    });

    it('should upload error when file not exists', function (done) {
      mm.http.request(/\/v1\/tfscom/, new Buffer(''), {});
      client.upload(logopath + 'not-exists', function (err, info) {
        should.exist(err);
        err.message.should.include('ENOENT');
        should.not.exist(info);
        done();
      });
    });

    it('should upload return wrong response', function (done) {
      mm.http.request(/\/v1\/tfscom/, new Buffer('{}'), {});
      client.upload(logopath, function (err, info) {
        should.exist(err);
        err.message.should.equal('TFS upload error');
        err.name.should.equal('TFSUploadError');
        should.not.exist(info);
        done();
      });
    });

    it('should upload timeout', function (done) {
      mm.http.request(/\/v1\/tfscom/, new Buffer('{}'), {}, 1000000);
      client.upload(logopath, function (err, info) {
        should.exist(err);
        err.message.should.equal('Request timeout for 500ms.');
        should.not.exist(info);
        done();
      }, 500);
    });
    
  });

  describe('refreshServers()', function () {

    var client;

    before(function () {
      client = tfs.createClient({
        appkey: 'tfscom',
        rootServer: '10.232.4.44:3800'
      });
    });

    it('should get servers list from rootServer', function (done) {
      client.once('servers', function (servers) {
        client.refreshCounter.should.equal(99); // getAppid()
        servers.length.should.above(0);
        client.removeAllListeners();
        done();
      });
      client.once('refreshError', function (err) {
        throw err;
      });

      client.refreshServers();
      client.refreshServers(); // no problem
    });

    it('should emit refreshError', function (done) {
      mm.http.requestError('/tfs.list', 'mock request() error');

      client.once('servers', function (servers) {
        throw new Error('should not emit this event');
      });
      client.once('refreshError', function (err) {
        should.exist(err);
        err.message.should.equal('mock request() error');
        done();
      });

      client.refreshServers();
    });

    it('should emit refreshError when statusCode !== 200', function (done) {
      mm.http.request('/tfs.list', '', { statusCode: 404 });

      client.once('servers', function (servers) {
        throw new Error('should not emit this event');
      });
      client.once('refreshError', function (err) {
        should.exist(err);
        err.message.should.equal('Http Response 404');
        done();
      });

      client.refreshServers();
    });

    it('should not emit any events when rootServer return empty datas', function (done) {
      mm.http.request('/tfs.list', '');
      client.once('servers', function (servers) {
        throw new Error('should not emit servers');
      });
      client.once('refreshError', function (err) {
        throw err;
      });
      client.refreshServers();
      setTimeout(done, 500);
    });

    it('should not emit any events when rootServer return wrong datas', function (done) {
      mm.http.request('/tfs.list', 'foo\nbar\n1\n\n2\n3\n12312312');
      client.once('servers', function (servers) {
        throw new Error('should not emit servers');
      });
      client.once('refreshError', function (err) {
        throw err;
      });
      client.refreshServers();
      setTimeout(done, 500);
    });

  });

  it('should emit ready event', function (done) {
    done = pedding(2, done);
    
    var c = tfs.createClient({
      appkey: 'tfscom',
      rootServer: '10.232.4.44:3800',
      imageServers: [
        'img01.daily.taobaocdn.net',
        'img02.daily.taobaocdn.net',
      ],
    });
    c.on('ready', function () {
      should.ok(c.appid);
      done();
    });
    // check queue
    c.uploadFile(logopath, 320, 'logo.png', function (err, info) {
      should.not.exist(err);
      info.should.have.keys('url', 'name', 'size');
      info.name.should.equal('L1/1/320/logo.png');
      info.url.should.include('http://');
      info.url.should.include('L1/1/320/logo.png');
      done();
    });
  });

  describe('createFile()', function () {

    it('should create 320/foobar.png', function (done) {
      done = pedding(2, done);

      tfsClient.createFile(320, 'foobar.png', function (err, success) {
        should.not.exist(err);
        should.ok(success);
        done();
      });

      tfsClient.createFile('320', 'foobar.png', function (err, success) {
        should.not.exist(err);
        should.ok(success);
        done();
      });
    });

    it('should return error when uid wrong', function (done) {
      tfsClient.createFile('wrongid', 'foobar.png', function (err, success) {
        should.exist(err);
        err.name.should.equal('TFSRequestError');
        err.message.should.equal('TFS request error, Http status 400');
        err.data.should.include('400 Bad Request');
        done();
      });
    });

  });

  describe('uploadFile()', function () {

    it('should uploadFile 320/logo.png', function (done) {
      tfsClient.uploadFile(logopath, 320, 'logo.png', function (err, info) {
        should.not.exist(err);
        info.should.have.keys('url', 'name', 'size');
        info.name.should.equal('L1/1/320/logo.png');
        info.url.should.include('http://');
        info.url.should.include('L1/1/320/logo.png');

        var p2 = path.join(__dirname, '1212.jpg');
        tfsClient.uploadFile(p2, 1212, 'head.jpg', function (err, info) {
          should.not.exist(err);
          info.should.have.keys('url', 'name', 'size');
          info.name.should.equal('L1/1/1212/head.jpg');
          info.url.should.include('http://');
          info.url.should.include('L1/1/1212/head.jpg');
          done();
        });
      });
    });

    it('should return error when uid wrong', function (done) {
      tfsClient.uploadFile(logopath, 'wrongid', 'logo.png', function (err, info) {
        should.exist(err);
        err.name.should.equal('TFSRequestError');
        err.message.should.equal('TFS request error, Http status 400');
        err.data.should.include('400 Bad Request');
        should.not.exist(info);
        done();
      });
    });

    it('should return error when file not exists', function (done) {
      tfsClient.uploadFile(logopath + 'not exists', 'wrongid', 'logo.png', function (err, info) {
        should.exist(err);
        err.message.should.include('ENOENT');
        should.not.exist(info);
        done();
      });
    });

  });

  describe('getAppid()', function () {
    it('should return appid', function (done) {
      tfsClient.getAppid(function (err, appid) {
        should.not.exist(err);
        appid.should.equal('1');
        done();
      });
    });

    it('should return mock error', function (done) {
      mm.http.requestError('/v2/tfscom/appid', 'get appid error');
      tfsClient.getAppid(function (err, appid) {
        should.exist(err);
        err.name.should.equal('MockHttpRequestError');
        err.message.should.equal('get appid error');
        should.not.exist(appid);
        done();
      });
    });

    it('should return mock empty response', function (done) {
      mm.http.request('/v2/tfscom/appid', '{}');
      tfsClient.getAppid(function (err, appid) {
        should.exist(err);
        err.name.should.equal('TFSRequestError');
        err.message.should.equal('GET /v2/tfscom/appid return appid is empty');
        should.not.exist(appid);
        done();
      });
    });
  });

  describe('removeFile()', function () {
    it('should remove not exists file success', function (done) {
      tfsClient.removeFile('320', 'noexists.jpg', function (err, success) {
        should.not.exist(err);
        should.ok(success);
        done();
      });
    });

    it('should return error when uid wrong', function (done) {
      tfsClient.removeFile('wronguid', 'noexists.jpg', function (err, success) {
        should.exist(err);
        err.message.should.equal('TFS request error, Http status 400')
        should.not.exist(success);
        done();
      });
    });

    it('should remove 320/logo.png', function (done) {
      tfsClient.uploadFile(logopath, 320, 'logo.png', function (err, info) {
        should.not.exist(err);
        info.should.have.keys('url', 'name', 'size');
        tfsClient.removeFile(320, 'logo.png', function (err, success) {
          should.not.exist(err);
          should.ok(success);
          done();
        });
      });
    });

  });

});

lib/command.js
/*!
 * tfs - lib/command.js
 * Copyright(c) 2012 fengmk2 <fengmk2@gmail.com>
 * MIT Licensed
 */

"use strict";

/**
 * Module dependencies.
 */

var path = require('path');
var util = require('util');

function Command(client, args) {
  this._args = args || {};
  this._client = client;
  this.ignoreStatus = null;
}

Command.prototype._getURL = function () {
  var url = '/v2/' + this._client.appkey + '/' + this._client.appid + '/' +
    this._args.uid + '/file/' + this._args.filename;
  return url;
};

Command.prototype.__defineGetter__('url', function () {
  return this._getURL();
});

Command.prototype.__defineGetter__('args', function () {
  var args = this._getArgs() || {};
  args.timeout = this._args.timeout;
  return args;
});

/**
 * v1: upload, remove, download, getMeta
 */

function UploadCommand(client, args) {
  UploadCommand.super_.call(this, client, args);
}
util.inherits(UploadCommand, Command);

UploadCommand.prototype._getURL = function () {
  var filepath = this._args.filepath;
  var pathname = '/v1/' + this._client.appkey;
  var extname = path.extname(filepath);
  if (extname) {
    pathname += '?suffix=' + extname + '&simple_name=1';
  }
  return pathname;
};

UploadCommand.prototype._getArgs = function () {
  return {
    type: 'POST',
    dataType: 'json',
    content: this._args.content,
  };
};

function RemoveCommand(client, args) {
  RemoveCommand.super_.call(this, client, args);
  this.ignoreStatus = 404;
}
util.inherits(RemoveCommand, Command);

RemoveCommand.prototype._getURL = function () {
  var url = '/v1/' + this._client.appkey + '/' + this._args.name;
  if (this._args.hide) {
    url += '?hide=' + this._args.hide; 
  }
  return url;
};

RemoveCommand.prototype._getArgs = function () {
  return {
    type: 'DELETE',
    headers: { 'Content-Length': 0 }
  };
};

function DownloadCommand(client, args) {
  DownloadCommand.super_.call(this, client, args);
}
util.inherits(DownloadCommand, Command);

DownloadCommand.prototype._getURL = function () {
  var url = '/v1/' + this._client.appkey + '/' + this._args.name;
  var offset = parseInt(this._args.offset, 10);
  var size = parseInt(this._args.size, 10);
  if (!isNaN(offset)) {
    url += '?offset=' + offset; 
    if (size > 0) {
      url += '&size=' + size;
    }
  }
  return url;
};

DownloadCommand.prototype._getArgs = function () {
  return {
    type: 'GET',
    writeStream: this._args.writeStream
  };
};

function GetMetaCommand(client, args) {
  GetMetaCommand.super_.call(this, client, args);
}
util.inherits(GetMetaCommand, Command);

GetMetaCommand.prototype._getURL = function () {
  var url = '/v1/' + this._client.appkey + '/metadata/' + this._args.name;
  var type = parseInt(this._args.type, 10);
  if (!isNaN(type)) {
    url += '?type=' + type;
  }
  return url;
};

GetMetaCommand.prototype._getArgs = function () {
  return {
    type: 'GET',
  };
};

/**
 * v2: createFile, uploadFile, removeFile, getFile
 */

function CreateFileCommand(client, args) {
  CreateFileCommand.super_.call(this, client, args);
  this.ignoreStatus = 409; // 409 Conflict, dir exists
}
util.inherits(CreateFileCommand, Command);

CreateFileCommand.prototype._getURL = function () {
  var url = CreateFileCommand.super_.prototype._getURL.call(this);
  return url + '?recursive=1';
};

CreateFileCommand.prototype._getArgs = function () {
  return {
    type: 'POST',
    headers: { 'Content-Length': 0 }
  };
};

function UploadFileCommand(client, args) {
  UploadFileCommand.super_.call(this, client, args);
}
util.inherits(UploadFileCommand, Command);

UploadFileCommand.prototype._getURL = function () {
  var url = UploadFileCommand.super_.prototype._getURL.call(this);
  if (this._args.offset) {
    url += '?offset=' + this._args.offset + '&size=' + this._args.content.length;
  }
  return url;
};

UploadFileCommand.prototype._getArgs = function () {
  return {
    type: 'PUT',
    content: this._args.content,
  };
};

function RemoveFileCommand(client, args) {
  RemoveFileCommand.super_.call(this, client, args);
  this.ignoreStatus = 404;
}
util.inherits(RemoveFileCommand, Command);

RemoveFileCommand.prototype._getArgs = function () {
  return {
    type: 'DELETE',
    headers: { 'Content-Length': 0 }
  };
};

// v1
exports.UploadCommand = UploadCommand;
exports.RemoveCommand = RemoveCommand;
exports.DownloadCommand = DownloadCommand;
exports.GetMetaCommand = GetMetaCommand;

// v2
exports.CreateFileCommand = CreateFileCommand;
exports.UploadFileCommand = UploadFileCommand;
exports.RemoveFileCommand = RemoveFileCommand;


lib/client.js
/*!
 * tfs - lib/client.js
 * Copyright(c) 2012 fengmk2 <fengmk2@gmail.com>
 * MIT Licensed
 */

"use strict";

/**
 * Module dependencies.
 */

var debug = require('debug')('tfs');
var path = require('path');
var fs = require('fs');
var util = require('util');
var EventEmitter = require('events').EventEmitter;
var urllib = require('urllib');
var eventproxy = require('eventproxy');
var command = require('./command');

/**
 * Create TFS RESTFul client.
 * 
 * @param {Object} options
 *  - {String} appkey
 *  - {String} [appLocation], default is 'tfscom'.
 *  - {Number} [uploadTimeout], upload max timeout, default is 60s.
 *  - {String} [rootServer], 'host:port' format, default is 'restful-store.vip.tbsite.net:3800'.
 *  - {Array} [imageServers], default is CDN online servers list.
 * @return {Client}
 */
function Client(options) {
  options = options || {};
  // http://baike.corp.taobao.com/index.php/CS_RD/tfs/http_server
  // apache服务器配置的应用路径，一般用于区分应用。如果你不需要自己独立的路径来区别，可以用默认的 "tfscom"
  this.appLocation = options.appLocation || 'tfscom';
  this.appkey = options.appkey;
  if (!this.appkey) {
    throw new TypeError('missing appkey');
  }

  this.rootServer = options.rootServer || 'restful-store.vip.tbsite.net:3800';
  var pair = this.rootServer.split(':');
  this.rootServerHost = pair[0];
  this.rootServerPort = pair[1] || 80;
  this.imageServers = options.imageServers || [
    'img1.tbcdn.cn',
    'img2.tbcdn.cn',
    'img3.tbcdn.cn',
    'img4.tbcdn.cn',
  ];
  this.imageServersIndex = 0;
  this.uploadTimeout = options.uploadTimeout || 60000;
  this.servers = [];
  this.serverIndex = 0;
  this.refreshCounter = 0;
  this.queue = [];
  this._init();
}

util.inherits(Client, EventEmitter);

Client.prototype._init = function () {
  var that = this;
  that._initTimer = setInterval(that.refreshServers.bind(that), 1000);

  that.once('servers', function () {
    // 初始化完成
    clearInterval(that._initTimer);
    that._initTimer = null;

    // get appid
    that.getAppid(function (err, appid) {
      if (err) {
        // TODO: user how to handle error?
        return that.emit('error', err);
      }
      that.appid = appid;

      // 处理队列
      for (var i = 0; i < that.queue.length; i++) {
        var task = that.queue[i];
        that._request.apply(that, task);
      }
      that.queue = [];
      that.emit('ready');
    });
  });
};

var SERVER_HOST_RE = /^\d+\.\d+\.\d+\.\d+/;

Client.prototype.refreshServers = function () {
  var that = this;
  if (that._freshing) {
    return;
  }
  that._freshing = true;
  var options = {
    host: this.rootServerHost,
    port: this.rootServerPort,
    path: '/tfs.list'
  };
  debug('GET /tfs.list');
  urllib.request(options, { timeout: 10000 }, function (err, data, res) {
    var status = res && res.statusCode;
    debug('GET /tfs.list %s', status);
    that._freshing = false;
    if (err) {
      err.url = options.path;
      return that.emit('refreshError', err);
    }

    if (status !== 200) {
      err = new Error('Http Response ' + res.statusCode);
      err.url = options.path;
      err.data = data && data.toString();
      return that.emit('refreshError', err);
    }

    if (data) {
      data = data.toString().trim();
    }

    var lines = data && data.split('\n') || [];
    if (lines.length < 2) {
      return;
    }

    // 应答内容如下，第一行为客户端下次从Webservice Root Server刷新TfsAgent Server的请求间隔，
    // 客户端发送的请求个数达到此数即应该重新从Webservice Root Server获取最新列表，
    // 其下按行存放了所有TfsAgent Server的地址：

    // 50
    // 10.232.4.41:3900
    // 10.232.4.42:3900
    // 10.232.4.43:3900
    var counter = parseInt(lines[0], 10) || 10;
    var servers = [];
    for (var i = 1; i < lines.length; i++) {
      var server = lines[i];
      if (!SERVER_HOST_RE.test(server)) {
        continue;
      }
      servers.push(server.split(':'));
    }
    if (servers.length === 0) {
      return;
    }

    that.refreshCounter = counter;
    that.serverIndex = 0;
    that.servers = servers;
    that.emit('servers', servers);
  });
};

Client.prototype._request = function (cmd, callback) {
  var that = this;
  if (that.servers.length === 0) {
    // add to queue
    return that.queue.push([ cmd, callback ]);
  }

  var server = that.servers[that.serverIndex++];
  if (that.serverIndex >= that.servers.length) {
    that.serverIndex = 0;
  }

  var pathname = cmd.url;
  var args = cmd.args || {};

  var options = {
    host: server[0],
    port: server[1],
    path: pathname,
  };
  args.timeout = args.timeout || that.uploadTimeout;

  debug('%s %s on %j, timeout %s', args.type, pathname, server, args.timeout);
  urllib.request(options, args, function (err, data, res) {
    var status = res && res.statusCode;
    var headers = res && res.headers;

    if (!err) {
      // success
      if (status === 200 || status === 201 || (cmd.ignoreStatus && status === cmd.ignoreStatus)) {
        debug('%s %j', status, headers);
        return callback(null, data);
      }

      // http non 200 error
      err = new Error('TFS request error, Http status ' + status);
      err.name = 'TFSRequestError';
    }
    
    err.status = status;
    err.headers = headers;
    err.url = server.join(':') + pathname;
    if (data) {
      if (Buffer.isBuffer(data)) {
        if (data.length > 1024) {
          data = data.slice(0, 1024);
        }
        data = data.toString();
      }
    }
    err.data = data;
    debug('%s %s: %s %s', args.type, pathname, status, err.message);
    callback(err);
  });

  // 判断是否重新获取 server 列表
  that.refreshCounter--;
  if (that.refreshCounter <= 0) {
    that.refreshServers();
  }
};

Client.prototype.getURL = function (pathname) {
  var imgServer = this.imageServers[this.imageServersIndex++];
  if (this.imageServersIndex >= this.imageServers.length) {
    this.imageServersIndex = 0;
  }
  return 'http://' + imgServer + '/' + pathname;
};

Client.prototype.getAppid = function (callback) {
  var pathname = '/v2/' + this.appkey + '/appid';
  var args = {
    type: 'GET',
    dataType: 'json'
  };
  this._request({ url: pathname, args: args }, function (err, info) {
    if (err) {
      return callback(err);
    }
    var appid = info && info.APP_ID;
    if (!appid) {
      err = new Error('GET ' + pathname + ' return appid is empty');
      err.name = 'TFSRequestError';
      return callback(err);
    }
    debug('appid: %s', appid);
    callback(null, appid);
  });
};

Client.prototype.createCommand = function (name, args) {
  var cmd = new command[name + 'Command'](this, args);
  cmd.__name = name;
  return cmd;
};

/**
 * v1
 */

/**
 * Upload a file.
 * @param {String} filepath
 * @param {Function(err, info)} callback
 *  - {Object} info
 *   - {String} name, tfs file name
 *   - {String} url, CDN url
 *   - {Number} size, file size
 * @param {Number} timeout, default is `client.uploadTimeout`.
 */
Client.prototype.upload = function (filepath, callback, timeout) {
  var that = this;
  fs.readFile(filepath, function (err, buffer) {
    if (err) {
      return callback(err);
    }

    var args = {
      filepath: filepath,
      timeout: timeout,
      content: buffer,
    };

    var cmd = that.createCommand('Upload', args);

    var size = buffer.length;
    debug('upload %d %s', size, filepath);
    that._request(cmd, function (err, data) {
      if (err) {
        return callback(err);
      }
      var name = data && data.TFS_FILE_NAME;
      if (!name) {
        // error
        err = new Error('TFS upload error');
        err.name = 'TFSUploadError';
        err.data = data;
        return callback(err);
      }

      var info = {
        name: name,
        size: size,
        url: that.getURL(that.appLocation + '/' + name)
      };
      debug('%j', info);
      callback(null, info);
    });
  });
};

/**
 * Remove a file by name.
 * 
 * @param {String} name
 * @param {Object} [options]
 *  - {Number|String} hide, 1: hidden, 0: show
 * @param {Function(err, success)} callback
 * @param {Number} timeout, default is `client.uploadTimeout`.
 */
Client.prototype.remove = function (name, options, callback, timeout) {
  if (typeof options === 'function') {
    timeout = callback;
    callback = options;
    options = null;
  }
  var args = {
    name: name,
    timeout: timeout,
  };
  if (options && typeof options.hide === 'number') {
    args.hide = String(options.hide);
  }

  var cmd = this.createCommand('Remove', args);
  this._request(cmd, function (err, data) {
    if (err) {
      return callback(err);
    }
    callback(null, true);
  });
};

/**
 * Download a file by name.
 * 
 * @param {String} name
 * @param {String|WriteStream} savefile, save file path or writable stream.
 * @param {Object} [options]
 *  - {Number} offset
 *  - {Number} size
 * @param {Function(err, success)} callback
 * @param {Number} timeout, default is `client.uploadTimeout`.
 */
Client.prototype.download = function (name, savefile, options, callback, timeout) {
  if (typeof options === 'function') {
    timeout = callback;
    callback = options;
    options = null;
  }
  var writeStream = savefile;
  if (typeof savefile === 'string') {
    writeStream = fs.createWriteStream(savefile);
  }
  var args = {
    name: name,
    timeout: timeout,
    writeStream: writeStream,
  };
  if (options) {
    args.offset = options.offset;
    args.size = options.size;
  }
  var cmd = this.createCommand('Download', args);
  this._request(cmd, function (err) {
    if (err) {
      return callback(err);
    }
    callback(null, true);
  });
};

/**
 * Get file meta.
 * @param {String} name
 * @param {Object} [options]
 *  - {Number} type, 0: normal mode, 1: force mode, get meta info even it was deleted.
 * @param {Function(err, meta)} callback
 * @param {Number} timeout
 */
Client.prototype.getMeta = function (name, options, callback, timeout) {
  if (typeof options === 'function') {
    timeout = callback;
    callback = options;
    options = null;
  }
  var args = {
    name: name,
    timeout: timeout,
  };
  if (options) {
    args.type = options.type;
  }
  var cmd = this.createCommand('GetMeta', args);
  this._request(cmd, function (err, data) {
    if (err) {
      return callback(err);
    }
    var meta = null;
    try {
      meta = JSON.parse(data);
    } catch (e) {
      err = e;
    }
    callback(err, meta);
  });
};

/**
 * v2
 */

/**
 * Create a file.
 * @param {String} uid, user id
 * @param {[type]} filename, e.g.: 'foo/bar.jpg', 'foo.png'.
 * @param {Function(err, success)} callback
 * @param {Number} [timeout], request timeout
 */
Client.prototype.createFile = function (uid, filename, callback, timeout) {
  var cmd = this.createCommand('CreateFile', {
    uid: uid,
    filename: filename,
    timeout: timeout
  });
  this._request(cmd, function (err, buffer) {
    if (err) {
      return callback(err);
    }
    callback(null, true);
  });
};

/**
 * Upload a file with custom filename.
 * 
 * @param {String} filepath
 * @param {String} uid, user id
 * @param {String} filename
 * @param {Object} [options]
 *  - {Number} offset
 * @param {Function(err, info)} callback
 * @param {Number} [timeout]
 */
Client.prototype.uploadFile = function (filepath, uid, filename, options, callback, timeout) {
  if (typeof options === 'function') {
    timeout = callback;
    callback = options;
    options = null;
  }

  var that = this;
  var ep = eventproxy.create('content', 'createFile', function (content, success) {
    var args = {
      uid: uid,
      content: content,
      filename: filename,
      timeout: timeout
    };

    if (options && options.offset) {
      args.offset = options.offset;
    }

    var cmd = that.createCommand('UploadFile', args);

    that._request(cmd, ep.done(function () {
      // http://image_host:port/L1/appid/userid/file_path
      // http://baike.corp.taobao.com/index.php/CS_RD/tfs/http_server
      var name = 'L1/' + that.appid + '/' + uid + '/' + filename;
      var info = {
        name: name,
        url: that.getURL(name),
        size: content.length
      };
      callback(null, info);
    }));
  });

  ep.fail(callback);

  fs.readFile(filepath, ep.done('content'));
  that.createFile(uid, filename, ep.done('createFile'), timeout);
};

/**
 * Remove a custom name file.
 * 
 * @param {String} uid
 * @param {String} filename
 * @param {Function(err, success)} callback
 * @param {Number} [timeout]
 */
Client.prototype.removeFile = function (uid, filename, callback, timeout) {
  var cmd = this.createCommand('RemoveFile', {
    uid: uid,
    filename: filename
  });

  this._request(cmd, function (err) {
    if (err) {
      return callback(err);
    }
    callback(null, true);
  });
};

exports.Client = Client;

exports.createClient = function (options) {
  return new Client(options);
};